pkg/rrd/rrd.go

```diff
@@ -0,0 +1,224 @@
+package rrd
+
+import (
+	"encoding/binary"
+	"math"
+	"time"
+
+	"github.com/boltdb/bolt"
+	"github.com/pkg/errors"
+)
+
+// RRD is a round robin database of fixed size which is specified on creation
+// RRD stores `counter` values. Which means values that can only go UP.
+// then it's easy to compute the increase of this counter over a given window
+// The database only keep history based on retention.
+type RRD interface {
+	// Slot returns the current window (slot) to store values.
+	Slot() (Slot, error)
+	// Counters, return all stored counters since the given time (since) until now.
+	Counters(since time.Time) (map[string]float64, error)
+	// Last returns the last reported value for a metric given the metric
+	// name
+	Last(key string) (value float64, ok bool, err error)
+}
+
+type Slot interface {
+	// Counter sets (or overrides) the current stored value for this key,
+	// with value
+	Counter(key string, value float64) error
+	// Key return the key of the slot which is the window timestamp
+	Key() uint64
+}
+
+type rrdBolt struct {
+	db        *bolt.DB
+	window    uint64
+	retention uint64
+}
+
+type rrdSlot struct {
+	db  *bolt.DB
+	key uint64
+}
+
+// NewRRDBolt creates a new rrd database that uses bolt as a storage. if window or retention are 0
+// the function will panic. If retnetion is smaller then window the function will panic.
+// retention and window must be multiple of 1 minute.
+func NewRRDBolt(path string, window time.Duration, retention time.Duration) (RRD, error) {
+	return newRRDBolt(path, window, retention)
+}
+
+func newRRDBolt(path string, window time.Duration, retention time.Duration) (*rrdBolt, error) {
+	window = (window / time.Minute) * time.Minute
+	retention = (retention / time.Minute) * time.Minute
+
+	if window == 0 {
+		panic("invalid window, can't be zero")
+	}
+	if retention == 0 {
+		panic("invalid retention, can't be zero")
+	}
+	if retention < window {
+		panic("retention can't be smaller than window")
+	}
+
+	db, err := bolt.Open(path, 0644, bolt.DefaultOptions)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open bolt db")
+	}
+
+	return &rrdBolt{
+		db:        db,
+		window:    uint64(window / time.Second),
+		retention: uint64(retention / time.Second),
+	}, nil
+}
+
+func (r *rrdBolt) retain(now uint64) error {
+	retain := now - r.retention
+	return r.db.Update(func(tx *bolt.Tx) error {
+		cur := tx.Cursor()
+
+		for k, _ := cur.First(); k != nil; k, _ = cur.Next() {
+			if len(k) != 8 {
+				continue // uknown key size
+			}
+
+			if lu64(k) <= retain {
+				// delete the bucket
+				// are we sure this is safe to do while iterating ?
+				_ = tx.DeleteBucket(k)
+			}
```

Comment:
Since the keys are ordered by size, if k > retain the loop can be ended without checking the rest of keys in the db?

Reply:
the keys are ordered sequentially actually, so when scanning the db. It will always be k (older timestamp) to latest timestamp